src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -108,20 +122,80 @@ Score(const size_t queryIndex, TreeType& referenceNode)
 
     // Calculate kernel value based on reference node centroid.
     if (tree::TreeTraits<TreeType>::FirstPointIsCentroid)
-    {
       kernelValue = EvaluateKernel(queryIndex, referenceNode.Point(0));
-    }
     else
-    {
-      kde::KDEStat& referenceStat = referenceNode.Stat();
       kernelValue = EvaluateKernel(queryPoint, referenceStat.Centroid());
-    }
 
     densities(queryIndex) += referenceNode.NumDescendants() * kernelValue;
 
     // Don't explore this tree branch.
     score = DBL_MAX;
   }
+  else if (monteCarlo &&
+           referenceNode.NumDescendants() >= MCAccessCoef * initialSampleSize &&
+           std::is_same<KernelType, kernel::GaussianKernel>::value)
+  {
+    // Calculate reference node depth.
+    if (referenceStat.Depth() == 0)
+      referenceStat.Depth() = CalculateDepth(referenceNode);
+
+    // Monte Carlo probabilistic estimation.
+    const double currentAlpha =
+        MCBeta / std::pow(2, referenceNode.Stat().Depth() + 1);
+    const boost::math::normal normalDist;
+    const double z =
+        std::abs(boost::math::quantile(normalDist, currentAlpha / 2));
+    const size_t numDesc = referenceNode.NumDescendants();
+    arma::vec sample;
+    size_t m = initialSampleSize;
+    double meanSample;
+    bool useMonteCarloPredictions = true;
+    while (m > 0)
+    {
+      const size_t oldSize = sample.size();
+      const size_t newSize = oldSize + m;
+
+      // Don't use probabilistic estimation if this is gonna take a close
+      // amount of computation to the exact calculation.
+      if (newSize >= MCBreakCoef * numDesc)
+      {
+        useMonteCarloPredictions = false;
+        break;
+      }
+
+      // Increase the sample size.
+      sample.resize(newSize);
+      for (size_t i = 0; i < m; ++i)
+      {
+        // Sample and evaluate random points from the reference node.
+        const size_t randomPoint = math::RandInt(0, numDesc);
+        sample(oldSize + i) =
+            EvaluateKernel(queryIndex, referenceNode.Descendant(randomPoint));
+      }
+      meanSample = arma::mean(sample);
+      const double stddev = arma::stddev(sample);
+      size_t mThresh =
+          std::ceil(z * stddev * (1 + relError) / (relError * meanSample));
+      mThresh *= mThresh;
```

Comment:
Correct me if I'm wrong, but shouldn't the `ceil()` come after `mThresh` is squared?

Reply:
You're right :)